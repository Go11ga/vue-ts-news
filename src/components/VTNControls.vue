<template>
    <div class="vtn-controls">
        <nav class="vtn-controls__nav">
            <router-link
                v-for="item in news.channels"
                :key="item.id"
                :to="`/${item.code}/1`"
                class="vtn-controls__link"
                :class="{ 'vtn-controls__link--active' : activeLink(item.code) }" 
            >
                {{ item.title }}
            </router-link>
        </nav>

        <div class="vtn-controls__item">
            <button
                v-for="item in modeBtns"
                :key="item.id"
                @click="setView(item.mode)"
                class="vtn-controls__btn"
                :class="{'vtn-controls__btn--active' : item.isActive}"
            >
                <svg class="vtn-controls__icon">
                    <use :xlink:href="`#icon_${item.mode}`"></use>
                </svg>
            </button>
        </div>
    </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import { mapState, mapMutations } from 'vuex'
import { IViewBtns } from '@/types/interfaces'

export default defineComponent({
    name: 'VTNControls',

    data() {
        return {
            modeBtns: [
                {id: 1, mode: 'grid', isActive: true},
                {id: 2, mode: 'list',  isActive: false}
            ] as Array<IViewBtns>
        }
    },

    computed: {
        ...mapState([
            'news'
        ])
    },

    methods: {
        ...mapMutations([
            'SET_VIEW_MODE'
        ]),

        setView(mode: string): void {
            let idx = this.modeBtns.findIndex(el => el.mode === mode)

            this.modeBtns.forEach((el) => el.isActive = false)
            this.modeBtns[idx].isActive = true

            localStorage.setItem('VTNViewMode', mode)

            this.SET_VIEW_MODE(mode)
        },

        activeLink(path: string): boolean {
            return path === this.news.activeChannel 
        }
    },

    created() {
        let viewMode = localStorage.getItem('VTNViewMode') || 'grid'

        this.setView(viewMode)
    }
})
</script>
