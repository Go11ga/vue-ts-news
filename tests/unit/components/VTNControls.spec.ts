import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'
import VTNControls from '@/components/VTNControls.vue'

describe('VTNControls', () => {
    let wrapper: any

    let setItem = jest.spyOn(Storage.prototype, 'setItem')
    let getItem = jest.spyOn(Storage.prototype, 'getItem').mockImplementation(() => 'grid')

    const state = () => {
        return {
            news: {
                channels: [
                    { id: 1, code: 'all', title: 'Все' },
                    { id: 2, code: 'lenta', title: 'Lenta' },
                    { id: 3, code: 'bbc-news', title: 'BBC' }
                ],
                activeChannel: 'lenta'
            }
        }
    }

    const mutations = {
        SET_VIEW_MODE: jest.fn()
    }

    const store = createStore({
        state,
        mutations
    })

    beforeEach(() => {
        wrapper = shallowMount(VTNControls, {
            global: {
                stubs: ['router-link'],
                plugins: [store]
            }
        })
    })

    afterEach(() => {
        wrapper.unmount()
    })

    it('dummy name test', () => {
        expect(wrapper.vm.$options.name).toMatch('VTNControls')
    })

    it('snapshot', () => {
        expect(wrapper.element).toMatchSnapshot()
    })

    it('router links', () => {
        expect(wrapper.findAll('.vtn-controls__link')).toHaveLength(3)
    })

    it('клик по кнопкам переключения вида', () => {
        wrapper.vm.setView = jest.fn()

        wrapper.find('.vtn-controls__btn').trigger('click')

        expect(wrapper.vm.setView).toHaveBeenCalled()
    })

    it('метод setView', async () => {
        wrapper.vm.setView('grid')

        await wrapper.vm.$nextTick()

        expect(wrapper.vm.modeBtns).toEqual(
            [
                { id: 1, mode: 'grid', isActive: true },
                { id: 2, mode: 'list', isActive: false }
            ]
        )
        expect(setItem).toHaveBeenCalled()
        expect(mutations.SET_VIEW_MODE).toHaveBeenCalled()
        expect(wrapper.findAll('.vtn-controls__btn')[0].classes()).toContain('vtn-controls__btn--active')

        wrapper.vm.setView('list')

        await wrapper.vm.$nextTick()

        expect(wrapper.vm.modeBtns).toEqual(
            [
                { id: 1, mode: 'grid', isActive: false },
                { id: 2, mode: 'list', isActive: true }
            ]
        )
        expect(setItem).toHaveBeenCalled()
        expect(mutations.SET_VIEW_MODE).toHaveBeenCalled()
        expect(wrapper.findAll('.vtn-controls__btn')[1].classes()).toContain('vtn-controls__btn--active')
    })

    it('метод activeLink', () => {
        expect(wrapper.vm.activeLink('lenta')).toBe(true)

        expect(wrapper.vm.activeLink('all')).toBe(false)
        expect(wrapper.vm.activeLink('bbc-news')).toBe(false)
    })

    it('хук created', async () => {
        expect(typeof VTNControls.created).toBe('function')

        wrapper.vm.setView = jest.fn()

        wrapper.vm.$options.created.call(wrapper.vm)

        expect(getItem).toHaveBeenCalled()
        expect(wrapper.vm.setView).toHaveBeenCalledWith('grid')
    })
})