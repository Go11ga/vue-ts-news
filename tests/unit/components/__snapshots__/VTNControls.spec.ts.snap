// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTNControls snapshot 1`] = `
<div
  class="vtn-controls"
>
  <nav
    class="vtn-controls__nav"
  >
    
    <router-link-stub
      class="vtn-controls__link"
      to="/all/1"
    />
    <router-link-stub
      class="vtn-controls__link vtn-controls__link--active"
      to="/lenta/1"
    />
    <router-link-stub
      class="vtn-controls__link"
      to="/bbc-news/1"
    />
    
  </nav>
  <div
    class="vtn-controls__item"
  >
    
    <button
      class="vtn-controls__btn vtn-controls__btn--active"
    >
      <svg
        class="vtn-controls__icon"
      >
        <use
          xlink:href="#icon_grid"
        />
      </svg>
    </button>
    <button
      class="vtn-controls__btn"
    >
      <svg
        class="vtn-controls__icon"
      >
        <use
          xlink:href="#icon_list"
        />
      </svg>
    </button>
    
  </div>
</div>
`;
